Cardano: Display account paths as account numbers

refi93 opened the pull request:
Motivation: Hide "raw" BIP32 path from users when displaying info about which account they are interacting with and instead display just the account index which is more user-friendly and also matches Trezor Connect experience where the account index is displayed as well.

related to (but not depending on) https://github.com/trezor/connect/pull/730

Changes:

* refactor warning prompts to warning-continue? prompts so users can actually reject the transaction if they find the content of the warning suspicious

* Update cardanoSignTransaction prompts for withdrawals to show account number instead of BIP32 path
Before:
![3_withdrawal](https://user-images.githubusercontent.com/4980147/105228022-566f7600-5b62-11eb-8352-d1983cd7ed98.png)

After:
![3withdrawal](https://user-images.githubusercontent.com/4980147/105227173-3ee3bd80-5b61-11eb-8e1a-e8e5ae8ed12a.png)


* Update prompts for staking certificates operations to show account number instead of BIP32 path
Before:
![2_certificate](https://user-images.githubusercontent.com/4980147/105227333-75213d00-5b61-11eb-9881-a3b1bd6c13b0.png)

After:
![2certificate](https://user-images.githubusercontent.com/4980147/105227349-7a7e8780-5b61-11eb-92e0-209dfd20a2e1.png)


* Update cardanoSignTransaction warning when change address staking account does not match the inputs:
Before:
![Screenshot 2020-12-17 23:05:48](https://user-images.githubusercontent.com/4980147/102550470-2abe2380-40be-11eb-869d-611702ee8743.png)
![1_different_staking_account](https://user-images.githubusercontent.com/4980147/105227230-591d9b80-5b61-11eb-815a-74909b7a3067.png)

After:
![Screenshot 2020-12-17 22:50:45](https://user-images.githubusercontent.com/4980147/102550422-19751700-40be-11eb-9250-43e8470be70b.png)
![1different_staking_account](https://user-images.githubusercontent.com/4980147/105227268-6175d680-5b61-11eb-9e4c-9b17297f6ec7.png)


* Update getAddress call warning when address staking account does not match the spending account
Before:
![4_stake_warning_address](https://user-images.githubusercontent.com/4980147/105227397-89fdd080-5b61-11eb-879f-4db28d5ca974.png)
![4_stake_warning_address_2](https://user-images.githubusercontent.com/4980147/105227401-8bc79400-5b61-11eb-909b-02723dbbf89d.png)


After:
![4address_stake_warning_1](https://user-images.githubusercontent.com/4980147/105227416-9124de80-5b61-11eb-8184-03369eba8fb2.png)
![4address_stake_warning_2](https://user-images.githubusercontent.com/4980147/105227421-92eea200-5b61-11eb-871f-690241f5e9ca.png)


Testing:
* device tests `pytest tests/device_tests -k "cardano"` pass

matejcik commented on core/src/apps/cardano/layout.py:
```diff
@@ -123,7 +123,7 @@ async def show_warning_tx_different_staking_account(
     page2 = Text("Confirm transaction", ui.ICON_SEND, ui.GREEN)
     page2.normal("Staking account:")
-    page2.bold(address_n_to_str(staking_account_path))
+    page2.bold("#%d" % to_account_number(staking_account_path))
```
what happens if a nonstandard path is used?

prusnak replied:
I think `to_account_number` should contain some sanity checks and return None or raise when they fail. In that case we can still show `address_n_to_str` instead.

refi93 replied:
we refactored it to `format_account_number()` and it enforces the path length to be at least 3, otherwise it raises a ValueError as that is an obviously malformed input, without doing further checks which should be handled outside of the logic to format the account number itself

tsusanka commented:
Ping @refi93, we have Freeze on Friday. In case you would like this included in the upcoming release, please let's finish this PR until then.

refi93 commented:
@tsusanka @matejcik code review fixes implemented, to sum it up:

* refactored the helper function for translating the account number to `format_account_number()`  which has the sole concern of rendering the account number appropriately and is not supposed to do more involved validation
* fixed warning prompts to include the option to cancel the operation for better UX as a reaction to a comment in another PR: https://github.com/trezor/trezor-firmware/pull/1390#discussion_r556427837

While testing, we also noticed a bug in the validation of (staking) paths which results in Trezor rejecting transactions that send the change to account numbers over 100 or perform withdrawals, so the user can only send all funds to an external address: https://github.com/vacuumlabs/trezor-firmware/issues/43 - we are going to address that in another PR which we will hopefully be able to deliver by Friday though we would appreciate if we got some more time and get it reviewed/merged even during the freeze as it is a bugfix to a relatively serious issue

refi93 commented on core/src/apps/cardano/layout.py:
```diff
@@ -440,13 +433,10 @@ async def show_warning_tx_network_unverifiable(ctx: wire.Context) -> None:
-        ),
-        button="Ok",
-    )
+    text = Text("Warning", ui.ICON_WRONG, ui.RED)
```
keeping the icon for consistency with previous UI but maybe there's a more suitable icon as technically there's nothing "wrong" per se

matejcik commented on core/src/apps/cardano/layout.py:
```diff
@@ -401,31 +401,24 @@ async def show_warning_address_foreign_staking_key(
+    page1.normal("Stake rights associated")
+    page1.normal("with this address do")
+    page1.normal("not match your")
+    page1.normal("account %s:" % format_account_number(account_path))
```
this still seems untrue when the path is non-standard, but that's presumably OK with the full derivation path shown anyway

matejcik reviewed (approved):
this seems good to go, please squash&rebase

matejcik commented:
Ah, no, let's not make the same mistake as before. Please run `pytest tests/device_tests --ui=record --ui-check-missing` to update the UI test fixtures first.

also in the future using the generated output of `--ui=test` is easier for us than custom before/after screenshots

note that 02881fd8 changes the way UI fixtures are named (taking the `description` field from the JSON), so make sure you rebase on top of that before regenerating

refi93 commented:
@matejcik ok thanks, good to know that, I will use the generated output next time. I rebased on top of latest master and regenerated UI test fixtures - is it expected though that the regeneration touched also non-cardano fixtures?

matejcik commented:
> is it expected though that the regeneration touched also non-cardano fixtures?

No.
Not sure what's going on, but first thing to try is, are you running the emulator with animations disabled?
Start `emu.py -a`, or use the make target `test_emu_ui_record` which launches its own emulator correctly

refi93 commented:
@matejcik thanks, the animations were apparently the culprit as now only the cardano-sepcific fixtures did change. I rebased the PR on top of the current master (2d4b91b52a4c1ba843cd077c584f1a7f04fc5119) as well. Maybe worth clarifying that animations should be disabled when updating the UI fixtures in Trezor's wiki: https://docs.trezor.io/trezor-firmware/tests/ui-tests.html

matejcik commented:
perfect, merging

> Maybe worth clarifying that animations should be disabled when updating the UI fixtures in Trezor's wiki: [docs.trezor.io/trezor-firmware/tests/ui-tests.html](https://docs.trezor.io/trezor-firmware/tests/ui-tests.html)

thanks for pointing this out, we'll update the doc

tsusanka commented:
Fixed 7b1e1b83b58d6281f91e30903b20a8fddc1b5714
